activerecord/lib/active_record/core.rb

```diff
@@ -19,11 +19,19 @@ module Core
       # retrieved on both a class and instance level by calling +logger+.
       class_attribute :logger, instance_writer: false
 
-      ##
-      # :singleton-method:
-      #
-      # Specifies the job used to destroy associations in the background
-      class_attribute :destroy_association_async_job, instance_writer: false, instance_predicate: false, default: false
+      class_attribute :_destroy_association_async_job, instance_writer: false, instance_predicate: false, default: "ActiveRecord::DestroyAssociationAsyncJob"
+      class_attribute :_destroy_association_async_job_class, instance_writer: false, instance_predicate: false
+
+      # The job class used to destroy associations in the background
+      def self.destroy_association_async_job
+        self._destroy_association_async_job_class ||= _destroy_association_async_job.respond_to?(:constantize) ? _destroy_association_async_job.constantize : _destroy_association_async_job
+      end
+
+      # Specifies the job used to destroy associations in the background, as a String
+      def self.destroy_association_async_job=(value)
+        self._destroy_association_async_job_class = nil
+        self._destroy_association_async_job = value
+      end
```

Comment:
If there is a typo in the configured class name, this will raise a vanilla `NameError`:

```ruby
class Post < ApplicationRecord
  self.destroy_association_async_job = "Oops"
  has_many :comments, dependent: :destroy_async
end
```

```console
> Post.first
.../activesupport/lib/active_support/inflector/methods.rb:279:in `constantize': uninitialized constant Oops (NameError)

      Object.const_get(camel_cased_word)
```

Which might be difficult to understand, depending on where the value is being set.

Also, is there a need for both the string and the resolved class?  Could we use a single `class_attribute` for both?  Something like:

```suggestion
      class_attribute :_destroy_association_async_job, instance_writer: false, instance_predicate: false, default: "ActiveRecord::DestroyAssociationAsyncJob"

      # The job class used to destroy associations in the background.
      def self.destroy_association_async_job
        if _destroy_association_async_job.is_a?(String)
          self._destroy_association_async_job = _destroy_association_async_job.constantize
        end
        _destroy_association_async_job
      rescue NameError => error
        if error.name == _destroy_association_async_job
          raise "Invalid class name for destroy_association_async_job: #{error.name.inspect}"
        else
          raise
        end
      end

      singleton_class.alias_method :destroy_association_async_job=, :_destroy_association_async_job=
```

Relatedly, I think the [error message in `check_dependent_options`](https://github.com/rails/rails/blob/c3c747547a85937e67bc62cfb9f466948fb631ad/activerecord/lib/active_record/associations/builder/association.rb#L131) should probably be updated because if `destroy_association_async_job` is `nil`, it will be for other reasons.

Reply:
@jonathanhefner yep, we can reduce it down to a single class_attribute if doing that repeated typecheck on String is ok. Probably? 

I'll take another look at the `check_dependent_options` too.